Add STM32 DMA consumer support for dts

cybertale opened the pull request:
Add generic DMA consumer and stm32 DMA consumer support in dts.

This is the dts part divided from #13364.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -599,6 +603,44 @@ def write_clocks(node):
+
+    # Client parameter syntax may vary with the DMA controller changes.
+    for dma_i, dma in enumerate(node.props["dmas"].val):
+        if "st,stm32-dma" in dma.controller.compats:
```
Could do this instead, to make it clearer that this is the only DMA controller that generates extra #defines:

```python
# Only this controller currently generates extra #defines
if "st,stm32-dma" not in dma.controller.compats:
    continue
```

A helper function could be used too.

cybertale replied:
Ok, thanks

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -599,6 +603,44 @@ def write_clocks(node):
+        if "st,stm32-dma" in dma.controller.compats:
+            dma.slot = dma.data["slot"]
+            dma.channel_config = dma.data["channel_config"]
+            dma.features = dma.data["features"]
```
Could use plain local variables (`slot`, `channel_config`, etc.) The edtlib objects should probably treated as immutable.

ulfalizer replied:
Would be nice to check that these keys exist ("channel_config", etc.) too, and exit with an error message otherwise. Check how the `err()` function is used.

The goal is for the new scripts to always fail with an error message, and never with a generic Python error.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -163,6 +164,9 @@ def write_props(node):
             continue
 
+        if prop.name == "dmas":
+            continue
```
Could add a comment to explain where it's handled instead.

ulfalizer commented on dts/bindings/dma/st,stm32-dma.yaml:
```diff
@@ -0,0 +1,37 @@
+title: STM32 DMA
+
+description: >
+    This is a representation of the STM32 DMA nodes
```
Not specific to this PR, but could use a shorter and more direct description like `description: STM32 DMA controller`.

The description shows up as a comment above any generated #defines.

ulfalizer commented on dts/bindings/dma/dma.yaml:
```diff
@@ -0,0 +1,22 @@
+#
```
Should probably call this file `dma-controller.yaml` instead. See the other `*-controller.yaml` files for a template.

ulfalizer commented on dts/bindings/dma/dma.yaml:
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2019, Song Qiang <songqiang1304521@gmail.com>
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Could use a more compact header style (other files have been changed the same way):

```
# Copyright (c) 2019, Song Qiang <songqiang1304521@gmail.com>
# SPDX-License-Identifier: Apache-2.0

...
```

ulfalizer commented on dts/bindings/dma/dma.yaml:
```diff
@@ -0,0 +1,22 @@
+title: DMA Base Structure
+
+description: >
+    This binding gives the base structures for all DMA devices
```
No need to have `title:` and `description:` keys in this file, since it's not a binding in itself and is just merged into actual bindings. See e.g. `gpio-controller.yaml` for a template.

ulfalizer commented on dts/bindings/dma/st,stm32-dma.yaml:
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2019, Song Qiang <songqiang1304521@gmail.com>
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Could use more compact header style:

```
# Copyright (c) 2019, Song Qiang <songqiang1304521@gmail.com>
# SPDX-License-Identifier: Apache-2.0

title: ...
```

ulfalizer commented on dts/bindings/dma/st,stm32-dma.yaml:
```diff
@@ -0,0 +1,37 @@
+# in the Linux kernel declared in
+# https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/stm32-dma.txt
+
+"dma-cells":
```
No need to put `""` around this key anymore.

ulfalizer reviewed (changes requested):
Some nits, mostly to make things consistent with some other changes.

I'm not familiar with the STM32 DMA controller, so other people should check the logic.

erwango commented on scripts/dts/gen_defines.py:
```diff
@@ -599,6 +603,44 @@ def write_clocks(node):
+
+    # Client parameter syntax may vary with the DMA controller changes.
+    for dma_i, dma in enumerate(node.props["dmas"].val):
+        if "st,stm32-dma" in dma.controller.compats:
```
Question to @ulfalizer for my understanding: I'm actually surprised we need to modify `gen_defines.py` script for compatible specific treatment. I would have expected yaml bindings help us to generate whatever is needed in a genric way.
Having defined `dma_cells` in bindings isn't it sufficient to generate matching defines?
How does it differ with gpio or clock cells?

ulfalizer replied:
Seems it's checking if particular bits in values in `(values)` in `dmas = <&ctrl (values)>` are set to determine which #defines to generate. No way to specify that in a binding. Would need to add a complete programming language to bindings to handle all the ways #defines might be generated.

Don't know anything about this controller, so don't know if it ought to be handled differently.

ulfalizer replied:
Another option would be to move the bit extraction stuff to the C code. Then the generic `phandle-array` handling could be used here.

cybertale replied:
The main reason why this is different from gpio or clock cells is that we want its dts parameter format to be consistent to the stm32 DMA consumer parameters format declared in Linux. If not, it would easy to declare more cells for each parameter and remove these decoding code.

ulfalizer replied:
Could have stuff like

    #define SOME_FLAG (DT_BLAH & 4)

in the fixup file too (or directly in the `.c` file... hope we can get rid of fixup files some day). That's still testable at compile time.

I could go either way here at least.

erwango replied:
> Another option would be to move the bit extraction stuff to the C code. Then the generic phandle-array handling could be used here.

I prefer this solution as well. 
I understand having this kind of "compatible" specific code could be handy, but it breaks the whole purpose of generic dts extraction script.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -599,6 +605,55 @@ def write_clocks(node):
+            # One common argument 'channel' and 3 private arguments for stm32
+            # dma controller.
+            if "slot" not in dma.data:
+                err("Missing 'slot' in DMA consumer node {}.", node.reg)
```
Probably want `{!r}` (the short `__repr__()` representation) and `dma.controller` instead of `node.reg`.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -599,6 +605,55 @@ def write_clocks(node):
+            # dma controller.
+            if "slot" not in dma.data:
+                err("Missing 'slot' in DMA consumer node {}.", node.reg)
+            if "channel_config" not in dma.data:
```
(Maybe this could be "channel-config" instead of "channel_config".)

ulfalizer commented:
> I prefer this solution as well.
> I understand having this kind of "compatible" specific code could be handy, but it breaks the whole purpose of generic dts extraction script.

Yeah, leaning towards this too.

@erwango Just because it might be hard to discover:

There is generic code for handling `phandle-array` in `write_props()` in `gen_defines.py`:

```python
        ...
        elif prop.type == "phandle-array":
            write_phandle_val_list(prop)
```

Normally you don't need to do any code changes to add new `phandle-array` properties.

cybertale commented:
> > I prefer this solution as well.
> > I understand having this kind of "compatible" specific code could be handy, but it breaks the whole purpose of generic dts extraction script.
> 
> Yeah, leaning towards this too.
> 
> @erwango Just because it might be hard to discover:
> 
> There is generic code for handling `phandle-array` in `write_props()` in `gen_defines.py`:
> 
> ```python
>         ...
>         elif prop.type == "phandle-array":
>             write_phandle_val_list(prop)
> ```
> 
> Normally you don't need to do any code changes to add new `phandle-array` properties.

Here we have to extract these parameters before the driver C code, or we would have to keep same code in every consumer driver. So must be written somewhere in the python dts extract script.

ulfalizer commented:
> Here we have to extract these parameters before the driver C code, or we would have to keep same code in every consumer driver. So must be written somewhere in the python dts extract script.

Could it be put in a common header file?

cybertale commented:
> > Here we have to extract these parameters before the driver C code, or we would have to keep same code in every consumer driver. So must be written somewhere in the python dts extract script.
> 
> Could it be put in a common header file?

You're right, that's better, it should be possible to put the extraction defines into a header file and remove these code in python. This way we would not need to change the gen_defines.py. 
Sorting and rebasing these commits are driving me crazy. I'll complete this  tomorrow with all other minor fixes.

ulfalizer commented:
> You're right, that's better, it should be possible to put the extraction defines into a header file and remove these code in python. This way we would not need to change the gen_defines.py.
> Sorting and rebasing these commits are driving me crazy. I'll complete this tomorrow with all other minor fixes.

The binding stuff looks good at least. Could put that into a separate PR maybe to get it in, if it's less messy.

b0661 commented:
>> The binding stuff looks good at least. Could put that into a separate PR maybe to get it in, if it's less messy.

Binding stuff should be extended by a binding for dma clients. This way clients only have to include dma-client.yaml:

Suggestion for dma-client.yaml:

```
#
title: DMA Client Base Structure

description: >
    This binding gives the base structures for all DMA clients.

properties:
    dmas:
      type: phandle-array
      required: true
      description: >
        List of one or more DMA specifiers, each consisting of
        - A phandle pointing to DMA controller node
        - A number of integer cells, as determined by the
        dma-cells property in the node referenced by phandle
        containing DMA controller specific information. This
        typically contains a DMA request line number or a
        channel number, but can contain any data that is
        required for configuring a channel..

    dma-names:
      type: string-array
      required: false
      description: >
        Contains one identifier string for each DMA specifier in
        the dmas property. The specific strings that can be used
        are defined in the binding of the DMA client device.
        Multiple DMA specifiers can be used to represent
        alternatives and in this case the dma-names for those
        DMA specifiers must be identical (see examples)
```

ulfalizer commented:
> Binding stuff should be extended by a binding for dma clients. This way clients only have to include dma-client.yaml:
> 
> Suggestion for dma-client.yaml:
> ...

Wonder if it could be called `dma-device.yaml`, to be consistent with other stuff. Could see the other `*-device.yaml` files for a template in that case.

avisconti commented on drivers/i2s/i2s_ll_stm32.c:
```diff
@@ -837,37 +857,46 @@ struct queue_item rx_1_ring_buf[CONFIG_I2S_STM32_RX_BLOCK_COUNT + 1];
+			.dma_slot = DT_I2S_1_DMA_SLOT_RX,
 			.channel_direction = PERIPHERAL_TO_MEMORY,
 			.source_data_size = 1,  /* 16bit default */
 			.dest_data_size = 1,    /* 16bit default */
```
Shouldn't source and dest data size be put equal to 2 as we said? This is number of bytes, so 16bit == 2 bytes.
An I think this commit should be squashed together with the commit for new DMA (e186b3ac), as they are inter-dependent (new DMA requires these changes, old DMA does not require these changes).

Do you agree?

cybertale replied:
I was making the changes. But before this, I want to write a macro like the one used in this driver https://github.com/cybertale/zephyr/blob/62dce55d73a2c2ca9ccd93e10c8d5ba927c38050/drivers/dma/dma_stm32.c#L533 to save some time since declaration code for I2S1-I2S5 has too many copies.

cybertale replied:
I was planning to do this in a new PR, but as you said only the new driver need this change, I will append them to this PR.

avisconti replied:
Sure, it looks a very good idea to have such a macro!!

cybertale commented:
> > Binding stuff should be extended by a binding for dma clients. This way clients only have to include dma-client.yaml:
> > Suggestion for dma-client.yaml:
> > ...
> 
> Wonder if it could be called `dma-device.yaml`, to be consistent with other stuff. Could see the other `*-device.yaml` files for a template in that case.
> 
> Edit: Hmm... seems `*-device.yaml` is only used for stuff on buses at the moment though, so do as you wish.

@b0661 Yes, it seems like to be consistent with others, *-device.yaml files are for other usage. The most similar driver to this would be the clock controller driver, and the corresponding description for 'dmas' you mentioned would be the 'clocks' in the base.yaml, which is directly declared in consumer nodes now.
I also think so that we should have a yaml file providing description for dmas for dma consumer nodes and clocks for the clock consumer nodes to import to avoid keeping multiple copies of the code. But that should be for another PR, as for this PR, I think we can keep it this way for now.

cybertale commented:
@ulfalizer I added a header file containing some macros to help extract data from `channel-config` and `features` now, thus no change will be needed to make to the gen_defines.py. It seems like not worth to create a new PR for the binding stuff now, so I'll keep them in this PR for now.

ulfalizer reviewed (approved):
No edtlib changes left in PR.

cybertale committed:
dt-bindings: add generic dma controller description support

Add a yaml file for help describing generic dma controllers.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
dt-bindings: add support for parsing stm32 dma consumer cells

Add support for parsing stm32 dma consumer cells, format of which
follows dma dts format declared in the Linux Kernel for the dma of
stm32:
https://git.kernel.org/pub/scm/linux/kernel/git/devicetree/devicetree-rebasing.git/plain/Bindings/dma/stm32-dma.txt

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
dt-bindings: arm: st: add dts support for DMA of some series of stm32

Add dts support for f0/f1/f2/f3/f4/f7/l0/l4 series STM32.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: code refactor

Declaration code for the I2S devices in the driver has too much
duplicate code. This commit uses a help macro to save some work
and some lines of code.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: change stm32 i2s driver to use two DMA controllers

The i2s driver assumes the tx channel and rx channel of dma are using
the same dma controller. This commit changes it to be able to use
different dma controllers.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: add some missing DMA parameters

src_addr_increment, dst_addr_increment, fifo_threshold and priority
are missing as parameters for configuring DMA. This commit adds them
to the driver.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
drivers: i2s: change the stm32 DMA in the driver to use dts

This commit moves DMA parameters previously hard coded in the driver
to the dts.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

cybertale committed:
dt-bindings: boards: apply dma dts changes to some boards

These boards using i2s in dts now have to declare the usage of dma in
dts, too.

Signed-off-by: Song Qiang <songqiang1304521@gmail.com>

galak commented:
Can we close this PR now?
